File: system-test/java/src/main/java/com/optivem/shop/testkit/port/when/WhenStage.java
package com.optivem.shop.testkit.port.when;

import com.optivem.shop.testkit.port.when.steps.WhenBrowseCoupons;
import com.optivem.shop.testkit.port.when.steps.WhenCancelOrder;
import com.optivem.shop.testkit.port.when.steps.WhenPlaceOrder;
import com.optivem.shop.testkit.port.when.steps.WhenPublishCoupon;
import com.optivem.shop.testkit.port.when.steps.WhenViewOrder;

public interface WhenStage {
    WhenPlaceOrder placeOrder();

    WhenCancelOrder cancelOrder();

    WhenViewOrder viewOrder();

    WhenPublishCoupon publishCoupon();

    WhenBrowseCoupons browseCoupons();
}

File: system-test/typescript/src/testkit/scenario/when/when-stage.ts
import { UseCaseContext } from '../../use-case-context.js';
import { AppContext } from '../app-context.js';
import { ScenarioContext } from '../scenario-context.js';
import { WhenPlaceOrder } from './when-place-order.js';
import { WhenCancelOrder } from './when-cancel-order.js';
import { WhenViewOrder } from './when-view-order.js';
import { WhenPublishCoupon } from './when-publish-coupon.js';
import { WhenBrowseCoupons } from './when-browse-coupons.js';

export class WhenStage {
  constructor(
    private readonly app: AppContext,
    private readonly ctx: ScenarioContext,
    private readonly useCaseContext: UseCaseContext,
  ) {}

  placeOrder(): WhenPlaceOrder {
    return new WhenPlaceOrder(this.app, this.ctx, this.useCaseContext);
  }

  cancelOrder(): WhenCancelOrder {
    return new WhenCancelOrder(this.app, this.ctx, this.useCaseContext);
  }

  viewOrder(): WhenViewOrder {
    return new WhenViewOrder(this.app, this.ctx, this.useCaseContext);
  }

  publishCoupon(): WhenPublishCoupon {
    return new WhenPublishCoupon(this.app, this.ctx, this.useCaseContext);
  }

  browseCoupons(): WhenBrowseCoupons {
    return new WhenBrowseCoupons(this.app, this.ctx, this.useCaseContext);
  }
}
